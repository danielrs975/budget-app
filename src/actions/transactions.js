/**
 * This script contains all the actions for
 * the transaction reducer
 * @author Daniel Rodriguez
 */
import uuid from 'uuid';

/**
 * The action to add a transaction
 * @param {*} transaction The transaction to add to the current state
 */
export const addTransaction = (transaction) => ({
    type: 'ADD_TRANSACTION',
    transaction,
});

/**
 * The action to remove a transaction
 * @param {} id the id of the transaction to remove
 */
export const removeTransaction = ({ id }) => ({
    type: 'REMOVE_TRANSACTION',
    id
});

/**
 * The action to change the info of a transaction
 * @param {*} id The id of the transaction to edit
 * @param {*} updates Changes
 */
export const editTransaction = (id, updates) => ({
    type: 'EDIT_TRANSACTION',
    id,
    updates
})