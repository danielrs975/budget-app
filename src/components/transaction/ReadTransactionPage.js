/**
 * This component is the page to see the information
 * about a transaction
 * @author Daniel Rodriguez
 */
import React from 'react'

export const ReadTransactionPage = (props) => (
    <div>
        
    </div>
)