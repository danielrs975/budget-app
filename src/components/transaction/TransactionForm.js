import React from 'react';

export class TransactionForm extends React.Component {
    /**
     * Handler of the type of transaction
     * field
     */
    onTypeChange = (e) => {
        const typeOfTransaction = e.target.value;
        this.setState(() => ({ typeOfTransaction }))
    }

    /**
     * Handler of the description
     * field
     */
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    /**
     * Handler of the date
     * field
     */
    onDateChange = (e) => {
        const date = e.target.value;
        this.setState(() => ({ date }));
    }
    
    /**
     * Handler of the amount
     * field
     */
    onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};

    // This field are activated when the type of transaction is "transfer"
    /**
     * Handler of the from account
     * field
     */
    onFromChange = (e) => {
        const from = e.target.value;
        this.setState(() => ({ from }));
    }

    /**
     * Handler of the to account
     * field
     */
    onToChange = (e) => {
        const to = e.target.value;
        this.setState(() => ({ to }))
    }
    //--------------------------------------------------------------------

    /**
     * Function to validate all the fields of the
     * form, for now only works with required fields
     */
    validateFields = () => {
        const errors = {};
        if (!this.state.amount) {
            errors['amount'] = 'The amount field is required';
        }

        if (!this.state.date) {
            errors['date'] = 'The date field is required';
        }
        
        if (this.state.typeOfTransaction === 'transfer') {
            if (!this.state.from) {
                errors['from'] = 'The from account is required';
            }
            if (!this.state.to) {
                errors['to'] = 'The to account is required';
            }
        }
        return errors;
    }

    /**
     * Handler of the submit
     * event
     */
    saveTransaction = (e) => {
        e.preventDefault();
        const errors = this.validateFields();
        const hasErrors = Object.keys(errors).length > 0
        if (!hasErrors) {
            this.setState(() => ({ errors }))
            let formValues = {
                typeOfTransaction: this.state.typeOfTransaction,
                description: this.state.description,
                date: this.state.date,
                amount: this.state.amount
            };
            if (this.state.typeOfTransaction === 'transfer') {
                formValues = {
                    ...formValues,
                    from: this.state.from,
                    to: this.state.to
                }
            }
    
            // After we send the information to redux
            console.log(formValues);
        } else {
            this.setState(() => ({ errors }));
        }
    }

    constructor(props) {
        super(props);
        this.state = {
            typeOfTransaction: 'purchase',
            description: '',
            date: '',
            from: '',
            to: '',
            amount: '',
            errors: {}
        };
    }
    
    render() {
        return (
            <div>
                <div className="content-container">
                    <form onSubmit={this.saveTransaction}>
                        <label>Type: </label>
                        <select value={this.state.typeOfTransaction} onChange={this.onTypeChange}>
                            <option value="purchase">Purchase</option>
                            <option value="transfer">Transfer</option>
                        </select>
                        <input 
                            type="text"
                            placeholder="Description"
                            value={this.state.description}
                            onChange={this.onDescriptionChange}
                        />
                        <input 
                            type="text"
                            placeholder="Date"
                            value={this.state.date}
                            onChange={this.onDateChange}
                        />
                        { this.state.errors.date && <span>{this.state.errors.date}</span> }
                        {/* This form is in case of transfer */}
                        { 
                            this.state.typeOfTransaction === 'transfer' 
                            && 
                            <div>
                                <label>From: </label>
                                <select value={this.state.from} onChange={this.onFromChange}>
                                    <option value="accountIid">Account I</option>
                                    <option value="accountIIid">Account II</option>
                                </select>
                                { this.state.errors.from && <span>{this.state.errors.from}</span> }
                                <label>To: </label>
                                <select value={this.state.to} onChange={this.onToChange}>
                                    <option value="accountIid">Account I</option>
                                    <option value="accountIIid">Account II</option>
                                </select>
                                { this.state.errors.to && <span>{this.state.errors.to}</span> }
                            </div>
                        }
                        <input 
                            type="text"
                            placeholder="Amount"
                            value={this.state.amount}
                            onChange={this.onAmountChange}
                        />
                        { this.state.errors.amount && <span>{this.state.errors.amount}</span> }
                        <button>Save transaction</button>
                    </form>
                </div>
            </div>
        );
    }
}

export default TransactionForm;