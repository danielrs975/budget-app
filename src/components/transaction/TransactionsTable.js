/**
 * Component that contain the table with all
 * the transactions
 */
import React from 'react'

const transactions = [
    {
        id: 1,
        amount: "12.00",
        description: "bread",
        typeOfTransaction: "purchase",
        date: "11/12/2020"
    },
    {
        id: 2,
        amount: "12.00",
        description: "to revolut",
        typeOfTransaction: "transfer",
        from: "revolut",
        to: "societe",
        date: "11/12/2020"
    }
]

export const TransactionsTable = (props) => (
    <div>
        <div className="content-container">
            { transactions.map((transaction) => (
                <div key={transaction.id}>
                    <h3>{transaction.description}</h3>
                    <p>Amount: {transaction.amount}</p>
                </div>
            )) }
        </div>
    </div>
)

export default TransactionsTable;