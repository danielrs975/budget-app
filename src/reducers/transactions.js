/**
 * This is the reducer for the transactions
 * information
 * @author Daniel Rodriguez
 */
export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_TRANSACTION':
            return [ ...state, action.transaction ];
        case 'REMOVE_TRANSACTION':
            return state.filter((transaction) => transaction.id !== action.id);
        case 'EDIT_TRANSACTION':
            return state.map((transaction) => {
                if (transaction.id === action.id) {
                    return { ...transaction, ...action.updates };
                }
                return transaction;
            })
        default:
            return state;
    }
};