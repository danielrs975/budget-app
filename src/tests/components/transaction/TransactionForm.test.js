/**
 * Tests for the transaction form
 * @author Daniel Rodriguez
 */
import React from 'react';
import { shallow } from 'enzyme';
import { TransactionForm } from '../../../components/transaction/TransactionForm';
import transactions from '../../fixtures/transactions';

let wrapper, onSubmit;
beforeEach(() => {
    onSubmit = jest.fn();
    wrapper = shallow(<TransactionForm onSubmit={onSubmit} />);
})

test('should render TransactionForm correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render TransactionForm correctly with values, purchase type', () => {
    wrapper = shallow(<TransactionForm transaction={transactions[0]} />)
    expect(wrapper).toMatchSnapshot();
})

test('should set type of transaction on select change', () => {
    const value = 'transfer';
    wrapper.find('select').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('typeOfTransaction')).toBe('transfer');
})

test('should render TransactionForm with transfer type transaction', () => {
    const value = 'transfer';
    wrapper.find('select').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper).toMatchSnapshot();
});

describe('Form when transfer type is selected', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<TransactionForm />);
        const value = 'transfer';
        wrapper.find('select').at(0).simulate('change', {
            target: { value }
        });
    });

    test('should set from account on select change', () => {
        const value = 'accountIid';
        wrapper.find('select').at(1).simulate('change', {
            target: { value }
        });
        expect(wrapper.state('from')).toBe(value);
    });

    test('should set to account on select change', () => {
        const value = 'accountIIid';
        wrapper.find('select').at(2).simulate('change', {
            target: { value }
        });
        expect(wrapper.state('to')).toBe(value);
    });

    test('should show errors in case of from and to account are empty', () => {
        wrapper.find('form').simulate('submit', {
            preventDefault: jest.fn()
        });
        expect(wrapper).toMatchSnapshot();
    })
})


test('should set descritpion on input change', () => {
    const value = 'Some description';
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('description')).toBe(value);
});

test('should set date on input change', () => {
    const value = '12/12/2020';
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('date')).toBe(value);
});

test('should show errors in case of date and amount empty', () => {
    wrapper.find('form').simulate('submit', {
        preventDefault: jest.fn()
    });
    expect(wrapper).toMatchSnapshot();
})

test('should set amount on input change', () => {
    const value = '22.1';
    wrapper.find('input').at(2).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('amount')).toBe(value);
})

test('should submit when the save transaction button is pressed', () => {
    wrapper = shallow(<TransactionForm transaction={transactions[0]} onSubmit={onSubmit} />);
    wrapper.find('form').simulate('submit', {
        preventDefault: jest.fn()
    });
    expect(onSubmit).toHaveBeenLastCalledWith(transactions[0]);
})