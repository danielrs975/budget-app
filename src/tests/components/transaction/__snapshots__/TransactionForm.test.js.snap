// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form when transfer type is selected should show errors in case of from and to account are empty 1`] = `
<div>
  <div
    className="content-container"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        Type: 
      </label>
      <select
        onChange={[Function]}
        value="transfer"
      >
        <option
          value="purchase"
        >
          Purchase
        </option>
        <option
          value="transfer"
        >
          Transfer
        </option>
      </select>
      <input
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        onChange={[Function]}
        placeholder="Date"
        type="text"
        value=""
      />
      <span>
        The date field is required
      </span>
      <div>
        <label>
          From: 
        </label>
        <select
          onChange={[Function]}
        >
          <option
            value="accountIid"
          >
            Account I
          </option>
          <option
            value="accountIIid"
          >
            Account II
          </option>
        </select>
        <span>
          The from account is required
        </span>
        <label>
          To: 
        </label>
        <select
          onChange={[Function]}
        >
          <option
            value="accountIid"
          >
            Account I
          </option>
          <option
            value="accountIIid"
          >
            Account II
          </option>
        </select>
        <span>
          The to account is required
        </span>
      </div>
      <input
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
      <span>
        The amount field is required
      </span>
      <button>
        Save transaction
      </button>
    </form>
  </div>
</div>
`;

exports[`should render TransactionForm correctly 1`] = `
<div>
  <div
    className="content-container"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        Type: 
      </label>
      <select
        onChange={[Function]}
        value="purchase"
      >
        <option
          value="purchase"
        >
          Purchase
        </option>
        <option
          value="transfer"
        >
          Transfer
        </option>
      </select>
      <input
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        onChange={[Function]}
        placeholder="Date"
        type="text"
        value=""
      />
      <input
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
      <button>
        Save transaction
      </button>
    </form>
  </div>
</div>
`;

exports[`should render TransactionForm correctly with values, purchase type 1`] = `
<div>
  <div
    className="content-container"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        Type: 
      </label>
      <select
        onChange={[Function]}
        value="purchase"
      >
        <option
          value="purchase"
        >
          Purchase
        </option>
        <option
          value="transfer"
        >
          Transfer
        </option>
      </select>
      <input
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value="bread"
      />
      <input
        onChange={[Function]}
        placeholder="Date"
        type="text"
        value={"1970-01-01T00:00:00.000Z"}
      />
      <input
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value={12000}
      />
      <button>
        Save transaction
      </button>
    </form>
  </div>
</div>
`;

exports[`should render TransactionForm with transfer type transaction 1`] = `
<div>
  <div
    className="content-container"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        Type: 
      </label>
      <select
        onChange={[Function]}
        value="transfer"
      >
        <option
          value="purchase"
        >
          Purchase
        </option>
        <option
          value="transfer"
        >
          Transfer
        </option>
      </select>
      <input
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        onChange={[Function]}
        placeholder="Date"
        type="text"
        value=""
      />
      <div>
        <label>
          From: 
        </label>
        <select
          onChange={[Function]}
        >
          <option
            value="accountIid"
          >
            Account I
          </option>
          <option
            value="accountIIid"
          >
            Account II
          </option>
        </select>
        <label>
          To: 
        </label>
        <select
          onChange={[Function]}
        >
          <option
            value="accountIid"
          >
            Account I
          </option>
          <option
            value="accountIIid"
          >
            Account II
          </option>
        </select>
      </div>
      <input
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
      <button>
        Save transaction
      </button>
    </form>
  </div>
</div>
`;

exports[`should show errors in case of date and amount empty 1`] = `
<div>
  <div
    className="content-container"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        Type: 
      </label>
      <select
        onChange={[Function]}
        value="purchase"
      >
        <option
          value="purchase"
        >
          Purchase
        </option>
        <option
          value="transfer"
        >
          Transfer
        </option>
      </select>
      <input
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        onChange={[Function]}
        placeholder="Date"
        type="text"
        value=""
      />
      <span>
        The date field is required
      </span>
      <input
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
      <span>
        The amount field is required
      </span>
      <button>
        Save transaction
      </button>
    </form>
  </div>
</div>
`;
